src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -337,4 +348,72 @@ void KafkaSource::calculateColumnPositions()
         physical_header.insert({physical_column.type->createColumn(), physical_column.type, physical_column.name});
     }
 }
+
+
+/// It basically initiate a checkpoint
+/// Since the checkpoint method is called in a different thread (CheckpointCoordinator)
+/// We nee make sure it is thread safe
+void KafkaSource::checkpoint(CheckpointContextPtr ckpt_ctx_)
+{
+    /// We assume the previous ckpt is already done
+    /// Use std::atomic<std::shared_ptr<CheckpointContext>>
+    assert(!ckpt_ctx.load(std::memory_order_relaxed));
+    ckpt_ctx = new CheckpointContext(*ckpt_ctx_);
+}
+
+/// 1) Generate a checkpoint barrier
+/// 2) Checkpoint the sequence number just before the barrier
+Chunk KafkaSource::doCheckpoint(CheckpointContextPtr ckpt_ctx_)
+{
+    /// Prepare checkpoint barrier chunk
+    auto result = header_chunk.clone();
+    auto chunk_ctx = std::make_shared<ChunkContext>();
```

Comment:
Would it make more sense to make a unique_ptr here instead?

Reply:
No, the `chunk_ctx` will be propagated downstream to sharing this